File: hostel_management_system_backend/Controllers/InteractionEventsController.cs
using Microsoft.AspNetCore.Mvc;

[ApiController]
[Route("api/[controller]")]
public sealed class InteractionEventsController : ControllerBase
{
    private readonly IInteractionEventsService _service;

    public InteractionEventsController(IInteractionEventsService service)
    {
        _service = service;
    }

    [HttpGet]
    public async Task<ActionResult<List<InteractionEventReadDto>>> GetAll(CancellationToken cancellationToken)
    {
        var events = await _service.GetAllAsync(cancellationToken);
        return Ok(events);
    }

    [HttpGet("{id:guid}")]
    public async Task<ActionResult<InteractionEventReadDto>> GetById(Guid id, CancellationToken cancellationToken)
    {
        var evt = await _service.GetByIdAsync(id, cancellationToken);
        return evt is null ? NotFound() : Ok(evt);
    }

    [HttpPost]
    public async Task<ActionResult<InteractionEventReadDto>> Create([FromBody] InteractionEventCreateDto dto, CancellationToken cancellationToken)
    {
        var created = await _service.CreateAsync(dto, cancellationToken);
        return CreatedAtAction(nameof(GetById), new { id = created.Id }, created);
    }

    [HttpPut("{id:guid}")]
    public async Task<ActionResult<InteractionEventReadDto>> Update(Guid id, [FromBody] InteractionEventUpdateDto dto, CancellationToken cancellationToken)
    {
        var updated = await _service.UpdateAsync(id, dto, cancellationToken);
        return updated is null ? NotFound() : Ok(updated);
    }

    [HttpDelete("{id:guid}")]
    public async Task<IActionResult> Delete(Guid id, CancellationToken cancellationToken)
    {
        var deleted = await _service.DeleteAsync(id, cancellationToken);
        return deleted ? NoContent() : NotFound();
    }
}

File: src/interaction-events/interaction-events.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseUUIDPipe,
  Post,
  Put,
} from '@nestjs/common';

import {
  InteractionEventCreateDto,
  InteractionEventReadDto,
  InteractionEventUpdateDto,
} from './dto/interaction-events.dto';
import { InteractionEventsService } from './interaction-events.service';

@Controller('interactionevents')
export class InteractionEventsController {
  constructor(private readonly interactionEventsService: InteractionEventsService) {}

  @Get()
  getAll(): Promise<InteractionEventReadDto[]> {
    return this.interactionEventsService.getAll();
  }

  @Get(':id')
  getById(@Param('id', ParseUUIDPipe) id: string): Promise<InteractionEventReadDto> {
    return this.interactionEventsService.getById(id);
  }

  @Post()
  create(@Body() dto: InteractionEventCreateDto): Promise<InteractionEventReadDto> {
    return this.interactionEventsService.create(dto);
  }

  @Put(':id')
  update(
    @Param('id', ParseUUIDPipe) id: string,
    @Body() dto: InteractionEventUpdateDto,
  ): Promise<InteractionEventReadDto> {
    return this.interactionEventsService.update(id, dto);
  }

  @Delete(':id')
  @HttpCode(HttpStatus.NO_CONTENT)
  async delete(@Param('id', ParseUUIDPipe) id: string) {
    await this.interactionEventsService.delete(id);
  }
}
